### DISCUSSION TITLE
GUI design

### DISCUSSION THREAD

**Discussion by aclist** [aclist/kbin-megamod#3] [Archived]
Initial implementation:

Expose a series of tabbed "pages" that contain semantic categories. Namespace these pages so that mods can be linked to discrete pages at the manifest level.

Current manifest POC: https://github.com/aclist/kbin-megamod/blob/main/manifest.json

The mod `foo` could be set to the page `inbox`, and when populating the GUI, this mod will get put in there. 

I haven't integrated this yet, but it is working. All of the metadata from the arrays can be moved into here and we can keep the root script clean. This also lets us store a lot of useful metadata that can be parsed later, and whitelisting of mods just entails modifying the manifest.json instead of changing the script itself. (The inline arrays were just being used to bootstrap the minimum viable functionality).

After the populate-by-manifest is working, I will bump to version 0.3.0. This should provide a cleaner framework for slotting in mods and making sure they go to the right place. Next step would be getting the tabs/pagination working correctly. (Buttons are just barebones code atm).

**Comment by artillect** [comment]
I like the direction you're going with the manifest.json. Like I said in #2, I think adding a field for the author's username would be good for attribution.

I think the tabs are definitely a good start, but I feel like a sidebar for the pages (like RES) would probably work better, at least once a lot of pages get added. This is what I had thrown together for the settings menu that I was working on before you released yours:

![image](https://github.com/aclist/kbin-megamod/assets/8906201/d71e3509-441b-415b-a4c3-0798cc66fbc1)

I'll admit it isn't really clear what's going on since there's only one sidebar option and one header, but each link in the sidebar scrolls the main view to the header with the same name. It's all done using anchor tags in the sidebar, and headers with the ID of the section.

**Reply by aclist** [reply]
Yes, I've added a field for author name in the manifest and experimented with putting it as a parenthetical remark next to the description, although it seems (slightly) verbose at first glance. Maybe not.

I have the fetch logic working and populating the fields from the JSON now, although I need to decouple it from the modal drawing so that the array can be created once on page load and stored, keeping the modal interaction snappy. GM_xmlHTTPrequest returns an async promise that needs to be parsed. A bit busy today so might not get around to finishing it.

I see, so basically the whole modal has a scrollbar inside of it? How many top-level "tabs" or "pages" is it realistic to have, assuming they are top-level categories and each one contains ~10 scripts (for example)? If the main view scrolls, seems like you could fit a lot in there without having a million pages. But I take your point that expansibility is important here or you are painting yourself into a corner.

Also keep in mind that it doesn't need to sit on the top quarter of the screen. If you fill the entire modal and have subsections that get redrawn, that's potentially a lot of room.

Hmm...what about blending tabs and anchors? So a "tab" represents a top-level section like "look and feel," and the anchors in the sidebar represent subsections to jump to so that you aren't lost in a sea of toggles.

E.g.

```
Page 1
   Sidebar 1 
       Option 1
       Option 2
   Sidebar 2 
      Option 1
      Option 2
```

```
Home page
   Hide options
      Hide sidebar
      Hide thumbnails
      Hide votes
   Layout options
      Full bleed layout
      Align right
Inbox
    Inbox page
       Format inbox into compact list
    Messages
       Hide avatar in message
       Show messages in reverse order
Threads
    Colors
       Colorize some thing
       Syntax highlighting
       Reply colors
   Hide options
      Hide badges
      Hide avatars
```

**Reply by artillect** [reply]
> I've added a field for author name in the manifest and experimented with putting it as a parenthetical remark next to the description, although it seems (slightly) verbose at first glance. Maybe not.

Sweet, I think it'll look less verbose as the number of options increases and the menu fills up. Once we add sub-options for mods and headers that'll help add some empty space back and make it a little easier on the eyes imo.

> I see, so basically the whole modal has a scrollbar inside of it? How many top-level "tabs" or "pages" is it realistic to have, assuming they are top-level categories and each one contains ~10 scripts (for example)? 

Yup, exactly. Theoretically you could have as many as you want if you make the sidebar scrollable/use an accordion like RES. (You could make the tab bar scrollable too but I think that's more cumbersome from a UX point of view).

Here's a screenshot of the RES settings UI for reference:

![a screenshot of the reddit enhancement suite UI](https://github.com/aclist/kbin-megamod/assets/8906201/4e75d4d1-420b-43c3-a2b9-678f211c9755)

> Also keep in mind that it doesn't need to sit on the top quarter of the screen. If you fill the entire modal and have subsections that get redrawn, that's potentially a lot of room.

I think I actually like it only covering the top part of the screen, since it gives you a preview of what the page will look like as you change your settings.

> Hmm...what about blending tabs and anchors? So a "tab" represents a top-level section like "look and feel," and the anchors in the sidebar represent subsections to jump to so that you aren't lost in a sea of toggles.

I'm worried that might overcomplicate things by adding too many layers, I think something like this would work well (maybe even breaking up the headers more granularly):

- Page 1
  - Header 1
  - Header 2
- Page 2
  - Header 1
  - Header 2

e.g:

- Home page
  - Hide elements
  - Layout
- Comments
  - Appearance
  - Layout
- etc.

**Reply by aclist** [reply]
> Sweet, I think it'll look less verbose as the number of options increases and the menu fills up. Once we add sub-options for mods and headers that'll help add some empty space back and make it a little easier on the eyes imo.

Take a look at it in the new version. Perhaps a tooltip or just a dedicated credits page would make this less noisy. Apropos of this, I'm leery of overengineering the UI itself because RES looks extremely busy to my eye. I'm in favor of loading it with features, but IMO readability should veer towards easy, so overloading too many widgets onscreen kinda of gets us back where kbin is already, I think. Striking a good balance between clearly demarcated sections and useful features is important.

> I think I actually like it only covering the top part of the screen, since it gives you a preview of what the page will look like as you change your settings.

I kind of like it, too; it's better than a weird semi-transparent modal blurring the screen, and you can see some of what is happening. Take a look at the new update, as I've added a docking button to let you pop the menu to the top or bottom of the screen, which could aid when looking at things on the navbar.

Now, to play devil's advocate with some rhetorical questions...


- If we want to preview the screen, why not open the whole menu in a new tab? Wouldn't this give you maximum preview functionality?
- What about tooltips that show a base64 encoded image depicting what the feature looks like when applied? Is this way too fancy? Probably.
- Aren't we sacrificing a lot of usable real estate here by truncating the size, and in turn making the small UI more fussy?
- Is scrollbars-within-the-viewport an anti-pattern? Is any form of pagination a detriment to usability?

Taking a page from RES seems appropriate, but we should also not reduplicate its conventions if they seem bad for our purposes. What's good for the goose may not be good for the gander. OTOH, at some screen resolution and scaling level the options are inevitably going to overflow for some people anyway, so accepting Y-overflow as a fact of life seems reasonable. (But if we can avoid using it as a crutch and keep most content "above the fold," so much the better.)

> I'm worried that might overcomplicate things by adding too many layers, I think something like this would work well (maybe even breaking up the headers more granularly):

I didn't fully undertand the example, will have to see a mockup. Didn't we describe the same thing? Top tier, sub tier, and options within.

Next order of business, I have deployed the manifest loading logic and defined a series of keys in there that make this more robust. Among them is version number, so atomic scripts could be tracked to a certain version that is decoupled from the megamod itself. 

The manifest also defines DOM element types like "checkbox," "textarea," "selector," etc, with "selector" being an array of values. Presumably, when building up the menus, the function adding the mods should pass this value and use a switch statement to create the right kind of field that the mod requests.

Next steps should be: 

- Simplify add-mod function and test alternate field types
- Also externalize the CSS so the megamod only contains business logic
- Manifest loading is an async process that triggers on page reload and contains a small race condition pitfall. Need to handle the promise response correctly and wait for the actual payload instead of assuming it has arrived. If it fails (such as GitHub being unreachable), need to prepare an alternate error message to populate the modal with.
- Minimize API usage: haven't thought much about this yet, but maybe bundling the requires into groups of scripts to minimize the number of requests that have to be made, but this kind of breaks the atomic script idea

Technical issue: I was able to get everything from the manifest out of the main file except the literal function name mappings (funcObj). Because JSON is returned as a string and there is no way to store a function name in JSON except stringwise, the stringwise function name needs to be interpolated into an actual function. I tried various invocations, such as `window['name']();`, etc., but they don't occupy the same scope, so it doesn't work. It seems like namespacing this with object literal mappings in the destination script should work (`megamod.func: func`), but I didn't think of a good way yet. For now, I just left the funcObj in the megamod, but it definitely has to go out of there. Basically, we need to take the entry point defined in the manifest and call that as a function after reading the variable.

**Reply by aclist** [reply]
WIP is on the [`patch-static-helpbox`](https://github.com/aclist/kbin-megamod/tree/patch-static-helpbox) branch. I still need to:

- [ ] Finish toggle functionality. Works but is not bound to any stateful action yet
- [x] Refactor inline `onclick` events into event listeners to make the sidebar/page events more robust (see below)
- [x] Start with the helpbox for mod at index 0  of the General page open

@artillect about the onclick events, this is brittle because:

- Necessitates writing the functional code inline in the DOM element
- Has a big scoping problem associated with it: since the helpbox contents are drawn dynamically after page load and the onclick function was being appended into the document body, it wasn't in the same scope as the functions in the userscript itself, so events taking place inside the helpbox (such as toggling) couldn't reach the `getSettings` or `applySettings` functions. This was fine when the toggles actually resided on a static part of the modal, but since everything is dynamic now, we need to set up the contents and hook listeners to them so that everything is in the same scope. (Plus, the inline scripts were a mess to manage). Event delegation ensures that the listener only needs to bind to the parent elements and will bubble through to children 

Basically, large blocks within backticks should be kept to a bare minimum, and ideally raw DOM element manipulation is kept as abstract as possible.

Update @ 1688118410: rewrote all of the dynamic logic and added event listeners. Toggle button is emitting stateful events now, just need to rewire it to the old code.

Next task: add more complex form fields and start integration testing mods. (Calling them mods as an internal name)

Question: added child breadcrumbs so that the header shows what mod you have opened, but this doesn't play well with the fact that we now start with the page open to whatever mod is in position 0 in the index. It means that the first time you open the modal, it looks like `Megamod Settings > General > Add mail link`, since we are already there. I'd rather just drop the child crumbs altogether (more generic) than write some fancy exclusion logic here, what do you think?
